openmls/src/ciphersuite/mod.rs

```diff
@@ -538,6 +549,203 @@ impl Signature {
+        fn encode_scalar(mut scalar: &[u8]) -> Result<Vec<u8>, SignatureError> {
+            // We first prepare the scalar DER encoded as a uint.
+            let mut scalar_uint: Vec<u8> = Vec::with_capacity(P256_SCALAR_LENGTH + 1);
+
+            // Remove prepending zeros
+            let mut msb = scalar
+                .read_u8()
+                .map_err(|_| SignatureError::EncodingError)?;
+            while msb == 0x00 {
+                msb = scalar
+                    .read_u8()
+                    .map_err(|_| SignatureError::EncodingError)?;
+            }
+
+            // If the most significant bit is 1, we have to prepend 0x00 to indicate
+            // that the integer is unsigned.
+            if msb > 0x7F {
+                scalar_uint
+                    .write_u8(0x00)
+                    .map_err(|_| SignatureError::EncodingError)?;
+            };
+
+            scalar_uint
+                .write_u8(msb)
+                .map_err(|_| SignatureError::EncodingError)?;
+            scalar_uint
+                .write_all(scalar)
+                .map_err(|_| SignatureError::EncodingError)?;
+
+            let mut encoded_scalar: Vec<u8> = Vec::with_capacity(P256_SCALAR_LENGTH + 3);
+            // The encoded scalar needs to start with integer tag.
+            encoded_scalar
+                .write_u8(INTEGER_TAG)
+                .map_err(|_| SignatureError::EncodingError)?;
+            // This is safe, as we know that r is at most 33 bytes long.
+            encoded_scalar
+                .write_u8(scalar_uint.len() as u8)
+                .map_err(|_| SignatureError::EncodingError)?;
+            encoded_scalar
+                .write_all(&scalar_uint)
+                .map_err(|_| SignatureError::EncodingError)?;
+
+            Ok(encoded_scalar)
+        }
+
+        // Check overall length
+        if raw_signature.len() != 2 * P256_SCALAR_LENGTH {
+            return Err(SignatureError::EncodingError);
+        }
+
+        // We DER encode the ECDSA signature as per spec, assuming that
+        // `sign` returns two concatenated values (r||s).
+        let r = raw_signature
+            .get(..P256_SCALAR_LENGTH)
+            .ok_or(SignatureError::EncodingError)?;
+        let s = raw_signature
+            .get(P256_SCALAR_LENGTH..2 * P256_SCALAR_LENGTH)
+            .ok_or(SignatureError::EncodingError)?;
+
+        let r_encoded = encode_scalar(&r)?;
+        let s_encoded = encode_scalar(&s)?;
+
+        let total_length = r_encoded.len() + s_encoded.len();
+        let mut encoded_signature: Vec<u8> = Vec::with_capacity(2 + 2 * (P256_SCALAR_LENGTH + 3));
+        encoded_signature
+            .write_u8(SEQUENCE_TAG)
+            .map_err(|_| SignatureError::EncodingError)?;
+        encoded_signature
+            .write_u8(total_length as u8)
+            .map_err(|_| SignatureError::EncodingError)?;
+        encoded_signature
+            .write_all(&r_encoded)
```

Comment:
So now you write the `r_encoded` again. This is the 3rd time I think. You can just pass the writer to `encode_scalar` to write it directly.